hw2/smart_args/smart_args_decorator.py

```diff
@@ -0,0 +1,66 @@
+from functools import update_wrapper
+from inspect import getfullargspec, FullArgSpec
+from typing import TypeVar, Generic, Callable, List, Tuple
+
+from hw2.smart_args.no_value import NoValue
+from hw2.smart_args.supported_magic_args import MagicArgument
+
+
+class MagicArgumentsMisuseError(TypeError):
+    """
+    The error that is raised when magic arguments are used where they should not be used.
+    """
+
+    pass
+
+
+def find_magic_kwargs(arg_spec: FullArgSpec) -> List[Tuple[str, MagicArgument]]:
+    """
+    Finds magic keyword arguments and their magic argument default values in the FullArgSpec
+    :param arg_spec: argument specification of the researched function
+    :return: list of tuples (keyword argument name, its magic argument default value)
+    """
+    kwonly_args = arg_spec.kwonlyargs
+    kwarg_defaults = arg_spec.kwonlydefaults or {}
+    return [
+        (kwarg_name, kwarg_defaults[kwarg_name])
+        for kwarg_name in kwonly_args
+        if kwarg_name in kwarg_defaults and isinstance(kwarg_defaults[kwarg_name], MagicArgument)
+    ]
+
+
+R = TypeVar("R")
+
+
+class SmartArgs(Generic[R]):
+    def __init__(self, func: Callable[..., R]):
+        args_data = getfullargspec(func)
+        self.func = func
+        self.positional_defaults = args_data.defaults
+        self.magic_kwargs = find_magic_kwargs(args_data)
+        self.check_magic_args_misuse()
+        update_wrapper(self, func)
+
+    def check_magic_args_misuse(self) -> None:
+        if self.positional_defaults is not None and any(
+            isinstance(default_value, MagicArgument) for default_value in self.positional_defaults
+        ):
+            raise MagicArgumentsMisuseError(
+                "Magic arguments can only be used as default values for keyword-only arguments"
+            )
+
+    def __call__(self, *args, **kwargs) -> R:
+        for kwarg_name, magic_default_value in self.magic_kwargs:
+            kwargs[kwarg_name] = magic_default_value(kwargs[kwarg_name] if kwarg_name in kwargs else NoValue)
+        return self.func(*args, **kwargs)
+
+
+def smart_args(func: Callable[..., R]) -> Callable[..., R]:
```

Comment:
Please, try to fix the following case:
```
HitCount = 0

def hit():
    global HitCount
    HitCount += 1
    print('hit', HitCount)
    return HitCount

print('define:')
@smart_args
def check_evaluation(*, x=hit(), y=hit, z=Evaluated(hit)):
    print('inside:')
    print(x, y(), z)

print('\nrun 1:')
check_evaluation(z=100)

print('\nrun 2:')
check_evaluation(x=100)

print('\nrun 3:')
check_evaluation()
```
The expected output is:
```
define:
hit 1

run 1:
inside:
hit 2
1 2 100

run 2:
hit 3
inside:
hit 4
100 4 3

run 3:
hit 5
inside:
hit 6
1 6 5
```
But in your case you will have a TypeError error

Reply:
Please refer to my last comment where I explain why this bug was present